### ISSUE TITLE
Appdaemon Pre-processor

### ISSUE THREAD

**Issue by turboc1208** [home-assistant/appdaemon#39]
The way you did the log files got me thinking.  It's almost a macro the way you did it.  How about a pre-processor that does simple string replacements into our code as it's loaded into appdaemon?  Here is why I am thinking this.  Continually typing 

`self.log("__function__ (__line__) here is my log entry")`
repeat 20 or 30 times while debugging which arguably is when you need the function and line numbers the most.  It's a pain.

even using a variable isn't much better.

`self.log(self.func + "("+self.line+") " + "here is my log entry") `
doesn't make it any better

but
  typing Dlog("here is my log entry")
and having it replaced at run time with
`self.log("__function__ (__line__) here is my log entry")`
would be really cool.

The problem with making Dlog a local function is that they any function and line references it puts in the log would reference Dlog as the function so it has to be some type of code substitution.  

Make sense??

**Comment by yawor** [comment]
This is not a macro or a preprocessor. It's a simple string replacement with values taken from Python's call frame stack.

If you want to change a default behaviour of `log` and `error` methods there should be no problem in overwriting them in your app's class:

```python
class MyApp(AppDaemon):

    def log(self, msg, level='INFO'):
        msg = self._sub_stack("__function__ (__line__) {}".format(msg))
        ha.log(self._logger, level, msg, self.name)

    def error(self, msg, level="WARNING"):
        msg = self._sub_stack("__function__ (__line__) {}".format(msg))
        ha.log(self._error, level, msg, self.name)
```

That way when you use anywhere else in your class `self.log('foo')` or `self.error('foo')` the function and line number will be automatically added. When you no longer need func and line no then just delete or comment out overwritten methods.

**Comment by turboc1208** [comment]
The string replacement is being done inside the AppDaemon libraries.  This is an example of a place where it could be helpful.  As I said in the message, continually typing all of the text to get the function and line number is a pain, it pushes the text further over in the code which makes it harder to read.  The problem with overriding it is that then there is another item on the stack that suddenly becomes the function that is getting it's name and line number put in for every log message.  The alternative is to not use the log function in AD and to write one of my own that gets the function and line number from inspect.  I don't like this option either because it doesn't use standard functionality of AD.

**Comment by yawor** [comment]
@turboc1208 have you actually tried the code I've posted in my previous comment? Overwriting a method doesn't add anything on the call stack during method call. I've used the same code that is in the base class. The only different is that the log message is being augmented with function and line number placeholders automatically.

As it is currently done in the AD, there's no other way to do that automatically and still use the built-in feature like you've said. But as long as you keep your call stack at the exact length even your own log function will work (you need to call `self._sub_stack` to replace placeholders directly from your log function).

To be honest, I don't really like how it is done and I'm going to prepare a PR which uses a log formatters instead of string replacement. There won't be any need to include something in the message itself.

**Comment by yawor** [comment]
As I've mentioned in the previous comment, I've created a PR #40 which removes the need for including `__module__`, `__line__` and `__function__` placeholders and moves this info to a format string.

**Comment by turboc1208** [comment]
I did not use your code, because I did that a couple of weeks ago, and have moved past it as not being the solution I was looking for.  The __module__, __line__ and __function__ were put in to make the inclusion of the module, line and function optional based on the original request to have them included in the log messages.  My first request was simply to have them included.  How they are included I really don't care, but given the option, I would choose less typing.
